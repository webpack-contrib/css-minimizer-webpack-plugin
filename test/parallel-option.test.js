import os from "os";

import { Worker } from "jest-worker";

import CssMinimizerPlugin from "../src/index";

import {
  compile,
  getCompiler,
  getErrors,
  getWarnings,
  readAssets,
} from "./helpers";

jest.mock("os", () => {
  const actualOs = jest.requireActual("os");

  const mocked = {
    cpus: jest.fn(() => {
      return { length: 4 };
    }),
  };

  return { ...actualOs, ...mocked };
});

// Based on https://github.com/facebook/jest/blob/edde20f75665c2b1e3c8937f758902b5cf28a7b4/packages/jest-runner/src/__tests__/test_runner.test.js
let workerTransform;
let workerEnd;

const ENABLE_WORKER_THREADS =
  typeof process.env.ENABLE_WORKER_THREADS !== "undefined"
    ? process.env.ENABLE_WORKER_THREADS === "true"
    : true;

jest.mock("jest-worker", () => {
  return {
    Worker: jest.fn().mockImplementation((workerPath) => {
      return {
        // eslint-disable-next-line global-require, import/no-dynamic-require
        transform: (workerTransform = jest.fn((data) =>
          // eslint-disable-next-line global-require, import/no-dynamic-require
          require(workerPath).transform(data)
        )),
        end: (workerEnd = jest.fn()),
        getStderr: jest.fn(),
        getStdout: jest.fn(),
      };
    }),
  };
});

const workerPath = require.resolve("../src/minify");

describe("parallel option", () => {
  let compiler;

  beforeEach(() => {
    jest.clearAllMocks();

    compiler = getCompiler({
      entry: {
        one: `${__dirname}/fixtures/entry.js`,
        two: `${__dirname}/fixtures/entry.js`,
        three: `${__dirname}/fixtures/entry.js`,
        four: `${__dirname}/fixtures/entry.js`,
      },
    });
  });

  it("should match snapshot when a value is not specify", async () => {
    new CssMinimizerPlugin().apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: os.cpus().length - 1,
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "false" value', async () => {
    new CssMinimizerPlugin({ parallel: false }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(0);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "true" value', async () => {
    new CssMinimizerPlugin({ parallel: true }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: Math.min(4, os.cpus().length - 1),
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "2" value', async () => {
    new CssMinimizerPlugin({ parallel: 2 }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: 2,
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "true" value when only one file passed', async () => {
    compiler = getCompiler({
      entry: `${__dirname}/fixtures/entry.js`,
    });

    new CssMinimizerPlugin({ parallel: true }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: Math.min(1, os.cpus().length - 1),
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "true" value and the number of files is less than the number of cores', async () => {
    const entries = {};

    for (let i = 0; i < os.cpus().length / 2; i++) {
      entries[`entry-${i}`] = `${__dirname}/fixtures/entry.js`;
    }

    compiler = getCompiler({ entry: entries });

    new CssMinimizerPlugin({ parallel: true }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: Math.min(Object.keys(entries).length, os.cpus().length - 1),
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "true" value and the number of files is same than the number of cores', async () => {
    const entries = {};

    for (let i = 0; i < os.cpus().length; i++) {
      entries[`entry-${i}`] = `${__dirname}/fixtures/entry.js`;
    }

    compiler = getCompiler({ entry: entries });

    new CssMinimizerPlugin({ parallel: true }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: Math.min(Object.keys(entries).length, os.cpus().length - 1),
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it('should match snapshot for the "true" value and the number of files is more than the number of cores', async () => {
    const entries = {};

    for (let i = 0; i < os.cpus().length * 2; i++) {
      entries[`entry-${i}`] = `${__dirname}/fixtures/entry.js`;
    }

    compiler = getCompiler({
      entry: {
        one: `${__dirname}/fixtures/entry.js`,
        two: `${__dirname}/fixtures/entry.js`,
        three: `${__dirname}/fixtures/entry.js`,
        four: `${__dirname}/fixtures/entry.js`,
        five: `${__dirname}/fixtures/entry.js`,
        six: `${__dirname}/fixtures/entry.js`,
        seven: `${__dirname}/fixtures/entry.js`,
        eight: `${__dirname}/fixtures/entry.js`,
      },
    });

    new CssMinimizerPlugin({ parallel: true }).apply(compiler);

    const stats = await compile(compiler);

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenLastCalledWith(workerPath, {
      enableWorkerThreads: ENABLE_WORKER_THREADS,
      numWorkers: Math.min(Object.keys(entries).length, os.cpus().length - 1),
    });
    expect(workerTransform).toHaveBeenCalledTimes(
      Object.keys(readAssets(compiler, stats, /\.css$/)).length
    );
    expect(workerEnd).toHaveBeenCalledTimes(1);

    expect(readAssets(compiler, stats, /\.css$/)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });
});
