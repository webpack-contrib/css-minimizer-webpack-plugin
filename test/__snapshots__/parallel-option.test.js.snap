// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parallel option should match snapshot for the "2" value: assets 1`] = `
{
  "four.css": "body{color:red}a{color:blue}",
  "one.css": "body{color:red}a{color:blue}",
  "three.css": "body{color:red}a{color:blue}",
  "two.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "2" value: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "2" value: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "false" value: assets 1`] = `
{
  "four.css": "body{color:red}a{color:blue}",
  "one.css": "body{color:red}a{color:blue}",
  "three.css": "body{color:red}a{color:blue}",
  "two.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "false" value: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "false" value: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is less than the number of cores: assets 1`] = `
{
  "entry-0.css": "body{color:red}a{color:blue}",
  "entry-1.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "true" value and the number of files is less than the number of cores: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is less than the number of cores: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is more than the number of cores: assets 1`] = `
{
  "eight.css": "body{color:red}a{color:blue}",
  "five.css": "body{color:red}a{color:blue}",
  "four.css": "body{color:red}a{color:blue}",
  "one.css": "body{color:red}a{color:blue}",
  "seven.css": "body{color:red}a{color:blue}",
  "six.css": "body{color:red}a{color:blue}",
  "three.css": "body{color:red}a{color:blue}",
  "two.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "true" value and the number of files is more than the number of cores: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is more than the number of cores: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is same than the number of cores: assets 1`] = `
{
  "entry-0.css": "body{color:red}a{color:blue}",
  "entry-1.css": "body{color:red}a{color:blue}",
  "entry-2.css": "body{color:red}a{color:blue}",
  "entry-3.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "true" value and the number of files is same than the number of cores: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value and the number of files is same than the number of cores: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value when only one file passed: assets 1`] = `
{
  "main.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "true" value when only one file passed: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value when only one file passed: warnings 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value: assets 1`] = `
{
  "four.css": "body{color:red}a{color:blue}",
  "one.css": "body{color:red}a{color:blue}",
  "three.css": "body{color:red}a{color:blue}",
  "two.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot for the "true" value: errors 1`] = `[]`;

exports[`parallel option should match snapshot for the "true" value: warnings 1`] = `[]`;

exports[`parallel option should match snapshot when a value is not specify: assets 1`] = `
{
  "four.css": "body{color:red}a{color:blue}",
  "one.css": "body{color:red}a{color:blue}",
  "three.css": "body{color:red}a{color:blue}",
  "two.css": "body{color:red}a{color:blue}",
}
`;

exports[`parallel option should match snapshot when a value is not specify: errors 1`] = `[]`;

exports[`parallel option should match snapshot when a value is not specify: warnings 1`] = `[]`;
